File: src/reviewiq/state.py
from __future__ import annotations

"""
ReviewIQ State Manager

Dual-backend state persistence:
  - CLI mode: local JSON files at .pr-review/reviews/pr-<N>.json
  - CI mode:  hidden comment on the PR (base64 JSON in HTML markers)

State tracks:
  - Findings with lifecycle (open -> resolved/wontfix/retracted)
  - Full conversation history (proper message pairs, not flat text)
  - Review rounds (which SHA was reviewed, what changed between rounds)
  - Per-finding discussion threads
"""

import json
import base64
import os
import re
import urllib.request
import urllib.error
from datetime import datetime, timezone
from pathlib import Path


# ── State Schema ─────────────────────────────────────────────────────────────

def new_state(pr_number: int, repo: str = "") -> dict:
    """Create a fresh state object for a PR."""
    return {
        "version": 2,
        "pr": {
            "number": pr_number,
            "repo": repo,
            "title": "",
            "author": "",
            "base_branch": "",
            "head_branch": "",
        },
        "review_rounds": [],
        "findings": {},
        "conversation": [],
        "summary": {
            "total_findings": 0,
            "open": 0,
            "resolved": 0,
            "wontfix": 0,
            "retracted": 0,
            "assessment": "PENDING",
            "last_reviewed_sha": "",
        },
    }


def new_finding(
    finding_id: int,
    title: str,
    severity: str,
    file: str,
    line: int,
    problem: str,
    impact: str,
    suggested_fix: str,
    fix_rationale: str,
    review_round: int,
) -> dict:
    """Create a new finding entry."""
    now = _now()
    return {
        "id": finding_id,
        "title": title,
        "severity": severity,
        "status": "open",
        "file": file,
        "line": line,
        "problem": problem,
        "impact": impact,
        "suggested_fix": suggested_fix,
        "fix_rationale": fix_rationale,
        "created_round": review_round,
        "created_at": now,
        "updated_at": now,
        "status_history": [
            {"status": "open", "round": review_round, "timestamp": now}
        ],
        "discussion": [],
    }


def new_review_round(
    round_number: int,
    head_sha: str,
    base_sha: str,
    event: str,
    files_reviewed: list[str],
) -> dict:
    """Create a new review round entry."""
    return {
        "round": round_number,
        "timestamp": _now(),
        "head_sha": head_sha,
        "base_sha": base_sha,
        "event": event,
        "files_reviewed": files_reviewed,
    }


# ── Finding Lifecycle ────────────────────────────────────────────────────────

VALID_STATUSES = {"open", "resolved", "partially_fixed", "wontfix", "retracted"}


def transition_finding(state: dict, finding_id: int, new_status: str, note: str = "", round_number: int = 0) -> dict:
    """Transition a finding to a new status with audit trail."""
    if new_status not in VALID_STATUSES:
        raise ValueError(f"Invalid status: {new_status}. Must be one of {VALID_STATUSES}")

    fid = str(finding_id)
    if fid not in state["findings"]:
        raise KeyError(f"Finding {finding_id} not found")

    finding = state["findings"][fid]
    finding["status"] = new_status
    finding["updated_at"] = _now()
    finding["status_history"].append({
        "status": new_status,
        "round": round_number,
        "timestamp": _now(),
        "note": note,
    })

    _recompute_summary(state)
    return state


def add_finding_discussion(state: dict, finding_id: int, role: str, content: str) -> dict:
    """Add a discussion message to a specific finding."""
    fid = str(finding_id)
    if fid not in state["findings"]:
        raise KeyError(f"Finding {finding_id} not found")

    state["findings"][fid]["discussion"].append({
        "role": role,
        "content": content,
        "timestamp": _now(),
    })
    return state


# ── Conversation History ─────────────────────────────────────────────────────

def add_message(state: dict, role: str, content: str, round_number: int = 0, metadata: dict | None = None) -> dict:
    """Append a message to the conversation history."""
    msg = {
        "role": role,
        "content": content,
        "round": round_number,
        "timestamp": _now(),
    }
    if metadata:
        msg["metadata"] = metadata
    state["conversation"].append(msg)
    return state


def get_conversation_for_llm(state: dict) -> list[dict]:
    """
    Convert conversation history to Claude API messages format.

    Returns list of {"role": "user"|"assistant", "content": "..."} dicts.
    System events become user messages prefixed with [SYSTEM].
    """
    messages = []
    for msg in state["conversation"]:
        if msg["role"] == "developer":
            messages.append({"role": "user", "content": msg["content"]})
        elif msg["role"] == "agent":
            messages.append({"role": "assistant", "content": msg["content"]})
        elif msg["role"] == "system":
            messages.append({"role": "user", "content": f"[SYSTEM EVENT] {msg['content']}"})

    # Claude requires alternating user/assistant. Merge consecutive same-role messages.
    if not messages:
        return messages

    merged = [messages[0]]
    for msg in messages[1:]:
        if msg["role"] == merged[-1]["role"]:
            merged[-1]["content"] += "\n\n" + msg["content"]
        else:
            merged.append(msg)

    # Claude requires first message to be user role
    if merged and merged[0]["role"] == "assistant":
        merged.insert(0, {"role": "user", "content": "[SYSTEM EVENT] Continuing previous review session."})

    return merged


# ── State Queries ────────────────────────────────────────────────────────────

def get_open_findings(state: dict) -> list[dict]:
    return [f for f in state["findings"].values() if f["status"] in ("open", "partially_fixed")]


def get_findings_by_status(state: dict, status: str) -> list[dict]:
    return [f for f in state["findings"].values() if f["status"] == status]


def get_latest_round(state: dict) -> dict | None:
    if not state["review_rounds"]:
        return None
    return state["review_rounds"][-1]


def get_finding(state: dict, finding_id: int) -> dict | None:
    return state["findings"].get(str(finding_id))


def get_state_summary_text(state: dict) -> str:
    """Generate a human-readable state summary for LLM context."""
    s = state["summary"]
    lines = [
        f"## Review State (Round {len(state['review_rounds'])})",
        f"- Total findings: {s['total_findings']}",
        f"- Open: {s['open']}",
        f"- Resolved: {s['resolved']}",
        f"- Won't fix: {s['wontfix']}",
        f"- Retracted: {s['retracted']}",
        f"- Assessment: {s['assessment']}",
        f"- Last reviewed SHA: {s['last_reviewed_sha'][:8] if s['last_reviewed_sha'] else 'none'}",
        "",
        "### Active Findings:",
    ]

    for f in get_open_findings(state):
        lines.append(
            f"  - **Finding {f['id']}** [{f['severity']}] ({f['status']}): "
            f"{f['title']} — `{f['file']}:{f['line']}`"
        )
        if f["discussion"]:
            lines.append(f"    Last discussion: {f['discussion'][-1]['content'][:100]}...")

    resolved = get_findings_by_status(state, "resolved")
    if resolved:
        lines.append("\n### Resolved Findings:")
        for f in resolved:
            last_note = ""
            for h in reversed(f["status_history"]):
                if h.get("note"):
                    last_note = f" — {h['note']}"
                    break
            lines.append(f"  - ~~Finding {f['id']}~~ [{f['severity']}]: {f['title']}{last_note}")

    return "\n".join(lines)


# ── Local File Backend ───────────────────────────────────────────────────────

def _state_dir() -> Path:
    return Path(".pr-review/reviews")


def _state_path(pr_number: int) -> Path:
    return _state_dir() / f"pr-{pr_number}.json"


def load_local(pr_number: int) -> dict | None:
    """Load state from local file. Returns None if no state exists."""
    path = _state_path(pr_number)
    if not path.exists():
        return None
    with open(path) as f:
        return json.load(f)


def save_local(state: dict) -> Path:
    """Save state to local file. Returns the path."""
    _state_dir().mkdir(parents=True, exist_ok=True)
    path = _state_path(state["pr"]["number"])
    with open(path, "w") as f:
        json.dump(state, f, indent=2)
    return path


# ── GitHub Comment Backend ───────────────────────────────────────────────────

STATE_MARKER_START = "<!-- REVIEWIQ_STATE_START -->"
STATE_MARKER_END = "<!-- REVIEWIQ_STATE_END -->"
STATE_COMMENT_HEADER = "<!-- REVIEWIQ_STATE_COMMENT -->"


def _gh_headers() -> dict:
    token = os.environ.get("GITHUB_TOKEN", "")
    return {
        "Authorization": f"token {token}",
        "Accept": "application/vnd.github.v3+json",
    }


def _gh_api(method: str, url: str, data: dict | None = None) -> dict | list | None:
    """Make a GitHub API call."""
    full_url = f"https://api.github.com{url}"
    body = json.dumps(data).encode() if data else None
    req = urllib.request.Request(full_url, data=body, method=method, headers=_gh_headers())
    if body:
        req.add_header("Content-Type", "application/json")
    try:
        with urllib.request.urlopen(req) as resp:
            if resp.status == 204:
                return None
            return json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"GitHub API error {e.code}: {e.read().decode()[:500]}") from e


def _encode_state(state: dict) -> str:
    """Encode state as base64 for embedding in a comment."""
    return base64.b64encode(json.dumps(state, separators=(",", ":")).encode()).decode()


def _decode_state(encoded: str) -> dict:
    """Decode base64 state from a comment."""
    return json.loads(base64.b64decode(encoded).decode())


def _find_state_comment(repo: str, pr_number: int) -> tuple[int | None, dict | None]:
    """Find the hidden state comment on a PR. Returns (comment_id, state) or (None, None)."""
    comments = _gh_api("GET", f"/repos/{repo}/issues/{pr_number}/comments")
    if not comments:
        return None, None

    for comment in comments:
        body = comment.get("body", "")
        if STATE_COMMENT_HEADER in body:
            match = re.search(
                rf"{re.escape(STATE_MARKER_START)}\n(.+?)\n{re.escape(STATE_MARKER_END)}",
                body,
                re.DOTALL,
            )
            if match:
                try:
                    state = _decode_state(match.group(1).strip())
                    return comment["id"], state
                except (json.JSONDecodeError, Exception):
                    pass
    return None, None


def load_remote(repo: str, pr_number: int) -> dict | None:
    """Load state from a hidden GitHub PR comment."""
    _, state = _find_state_comment(repo, pr_number)
    return state


def save_remote(state: dict) -> None:
    """Save state to a hidden GitHub PR comment (create or update)."""
    repo = state["pr"]["repo"]
    pr_number = state["pr"]["number"]
    encoded = _encode_state(state)

    # Build a minimal visible summary + hidden state blob
    summary = state["summary"]
    body = f"""{STATE_COMMENT_HEADER}
<details>
<summary>ReviewIQ State (Round {len(state['review_rounds'])}) — {summary['open']} open, {summary['resolved']} resolved</summary>

| Metric | Count |
|--------|-------|
| Total findings | {summary['total_findings']} |
| Open | {summary['open']} |
| Resolved | {summary['resolved']} |
| Won't fix | {summary['wontfix']} |
| Retracted | {summary['retracted']} |
| Assessment | {summary['assessment']} |

</details>

{STATE_MARKER_START}
{encoded}
{STATE_MARKER_END}"""

    comment_id, _ = _find_state_comment(repo, pr_number)
    if comment_id:
        _gh_api("PATCH", f"/repos/{repo}/issues/comments/{comment_id}", {"body": body})
    else:
        _gh_api("POST", f"/repos/{repo}/issues/{pr_number}/comments", {"body": body})


# ── Unified Load/Save ────────────────────────────────────────────────────────

def load(pr_number: int, repo: str = "", prefer: str = "auto") -> dict:
    """
    Load state from the best available backend.

    prefer:
      "local"  — local file only
      "remote" — GitHub comment only
      "auto"   — try local first, then remote, then create new
    """
    state = None

    if prefer in ("local", "auto"):
        state = load_local(pr_number)

    if state is None and prefer in ("remote", "auto") and repo:
        state = load_remote(repo, pr_number)

    if state is None:
        state = new_state(pr_number, repo)

    return state


def save(state: dict, targets: str = "auto") -> None:
    """
    Save state to the specified backends.

    targets:
      "local"  — local file only
      "remote" — GitHub comment only
      "both"   — both backends
      "auto"   — local always, remote if repo is set and GITHUB_TOKEN exists
    """
    if targets in ("local", "both", "auto"):
        save_local(state)

    if targets in ("remote", "both"):
        save_remote(state)
    elif targets == "auto" and state["pr"]["repo"] and os.environ.get("GITHUB_TOKEN"):
        save_remote(state)


# ── Helpers ──────────────────────────────────────────────────────────────────

def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _recompute_summary(state: dict) -> None:
    """Recompute the summary counts from findings."""
    findings = state["findings"].values()
    state["summary"]["total_findings"] = len(state["findings"])
    state["summary"]["open"] = sum(1 for f in findings if f["status"] in ("open", "partially_fixed"))
    state["summary"]["resolved"] = sum(1 for f in findings if f["status"] == "resolved")
    state["summary"]["wontfix"] = sum(1 for f in findings if f["status"] == "wontfix")
    state["summary"]["retracted"] = sum(1 for f in findings if f["status"] == "retracted")

    # Auto-assess
    open_count = state["summary"]["open"]
    has_critical = any(
        f["severity"] == "CRITICAL" and f["status"] in ("open", "partially_fixed")
        for f in findings
    )
    if open_count == 0 and state["summary"]["total_findings"] > 0:
        state["summary"]["assessment"] = "APPROVE"
    elif has_critical:
        state["summary"]["assessment"] = "REQUEST CHANGES"
    elif open_count > 0:
        state["summary"]["assessment"] = "REQUEST CHANGES"
    else:
        state["summary"]["assessment"] = "PENDING"

File: internal/state/state.go
// Package state manages ReviewIQ review state via GitHub PR comments.
package state

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/Sanmanchekar/reviewiq/internal/github"
)

// ── Types ───────────────────────────────────────────────────────────────────

type PRMetadata struct {
	Number     int    `json:"number"`
	Repo       string `json:"repo"`
	Title      string `json:"title"`
	Author     string `json:"author"`
	BaseBranch string `json:"base_branch"`
	HeadBranch string `json:"head_branch"`
}

type ReviewRound struct {
	Round         int      `json:"round"`
	Timestamp     string   `json:"timestamp"`
	HeadSHA       string   `json:"head_sha"`
	BaseSHA       string   `json:"base_sha"`
	Event         string   `json:"event"`
	FilesReviewed []string `json:"files_reviewed"`
}

type StatusEntry struct {
	Status    string `json:"status"`
	Round     int    `json:"round"`
	Timestamp string `json:"timestamp"`
	Note      string `json:"note,omitempty"`
}

type DiscussionMsg struct {
	Role      string `json:"role"`
	Content   string `json:"content"`
	Timestamp string `json:"timestamp"`
}

type Finding struct {
	ID            int             `json:"id"`
	Title         string          `json:"title"`
	Severity      string          `json:"severity"`
	Status        string          `json:"status"`
	File          string          `json:"file"`
	Line          int             `json:"line"`
	Problem       string          `json:"problem"`
	Impact        string          `json:"impact"`
	SuggestedFix  string          `json:"suggested_fix"`
	FixRationale  string          `json:"fix_rationale"`
	CreatedRound  int             `json:"created_round"`
	CreatedAt     string          `json:"created_at"`
	UpdatedAt     string          `json:"updated_at"`
	StatusHistory []StatusEntry   `json:"status_history"`
	Discussion    []DiscussionMsg `json:"discussion"`
}

type ConversationMsg struct {
	Role      string            `json:"role"`
	Content   string            `json:"content"`
	Round     int               `json:"round"`
	Timestamp string            `json:"timestamp"`
	Metadata  map[string]string `json:"metadata,omitempty"`
}

type Summary struct {
	TotalFindings  int    `json:"total_findings"`
	Open           int    `json:"open"`
	Resolved       int    `json:"resolved"`
	Wontfix        int    `json:"wontfix"`
	Retracted      int    `json:"retracted"`
	Assessment     string `json:"assessment"`
	LastReviewedSHA string `json:"last_reviewed_sha"`
}

type ReviewState struct {
	Version      int                `json:"version"`
	PR           PRMetadata         `json:"pr"`
	ReviewRounds []ReviewRound      `json:"review_rounds"`
	Findings     map[string]Finding `json:"findings"`
	Conversation []ConversationMsg  `json:"conversation"`
	Summary      Summary            `json:"summary"`
}

type LLMMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// ── Constructors ────────────────────────────────────────────────────────────

func NewState(prNumber int, repo string) *ReviewState {
	return &ReviewState{
		Version:      2,
		PR:           PRMetadata{Number: prNumber, Repo: repo},
		ReviewRounds: []ReviewRound{},
		Findings:     make(map[string]Finding),
		Conversation: []ConversationMsg{},
		Summary: Summary{
			Assessment: "PENDING",
		},
	}
}

func NewFinding(id int, title, severity, file string, line int, problem, impact, fix, rationale string, round int) Finding {
	now := nowUTC()
	return Finding{
		ID: id, Title: title, Severity: severity, Status: "open",
		File: file, Line: line, Problem: problem, Impact: impact,
		SuggestedFix: fix, FixRationale: rationale,
		CreatedRound: round, CreatedAt: now, UpdatedAt: now,
		StatusHistory: []StatusEntry{{Status: "open", Round: round, Timestamp: now}},
		Discussion:    []DiscussionMsg{},
	}
}

func NewReviewRound(round int, headSHA, baseSHA, event string, files []string) ReviewRound {
	return ReviewRound{
		Round: round, Timestamp: nowUTC(),
		HeadSHA: headSHA, BaseSHA: baseSHA,
		Event: event, FilesReviewed: files,
	}
}

// ── Finding Lifecycle ───────────────────────────────────────────────────────

var ValidStatuses = map[string]bool{
	"open": true, "resolved": true, "partially_fixed": true,
	"wontfix": true, "retracted": true,
}

func (s *ReviewState) TransitionFinding(findingID int, newStatus, note string, round int) error {
	if !ValidStatuses[newStatus] {
		return fmt.Errorf("invalid status: %s", newStatus)
	}
	fid := fmt.Sprintf("%d", findingID)
	f, ok := s.Findings[fid]
	if !ok {
		return fmt.Errorf("finding %d not found", findingID)
	}
	f.Status = newStatus
	f.UpdatedAt = nowUTC()
	f.StatusHistory = append(f.StatusHistory, StatusEntry{
		Status: newStatus, Round: round, Timestamp: nowUTC(), Note: note,
	})
	s.Findings[fid] = f
	s.RecomputeSummary()
	return nil
}

func (s *ReviewState) AddFindingDiscussion(findingID int, role, content string) error {
	fid := fmt.Sprintf("%d", findingID)
	f, ok := s.Findings[fid]
	if !ok {
		return fmt.Errorf("finding %d not found", findingID)
	}
	f.Discussion = append(f.Discussion, DiscussionMsg{
		Role: role, Content: content, Timestamp: nowUTC(),
	})
	s.Findings[fid] = f
	return nil
}

// ── Conversation ────────────────────────────────────────────────────────────

func (s *ReviewState) AddMessage(role, content string, round int, metadata map[string]string) {
	s.Conversation = append(s.Conversation, ConversationMsg{
		Role: role, Content: content, Round: round,
		Timestamp: nowUTC(), Metadata: metadata,
	})
}

func (s *ReviewState) GetConversationForLLM() []LLMMessage {
	var msgs []LLMMessage
	for _, m := range s.Conversation {
		switch m.Role {
		case "developer":
			msgs = append(msgs, LLMMessage{Role: "user", Content: m.Content})
		case "agent":
			msgs = append(msgs, LLMMessage{Role: "assistant", Content: m.Content})
		case "system":
			msgs = append(msgs, LLMMessage{Role: "user", Content: "[SYSTEM EVENT] " + m.Content})
		}
	}
	if len(msgs) == 0 {
		return msgs
	}
	// Merge consecutive same-role messages
	merged := []LLMMessage{msgs[0]}
	for _, m := range msgs[1:] {
		last := &merged[len(merged)-1]
		if m.Role == last.Role {
			last.Content += "\n\n" + m.Content
		} else {
			merged = append(merged, m)
		}
	}
	// First message must be user
	if merged[0].Role == "assistant" {
		merged = append([]LLMMessage{{Role: "user", Content: "[SYSTEM EVENT] Continuing previous review session."}}, merged...)
	}
	return merged
}

// ── Queries ─────────────────────────────────────────────────────────────────

func (s *ReviewState) OpenFindings() []Finding {
	var out []Finding
	for _, f := range s.Findings {
		if f.Status == "open" || f.Status == "partially_fixed" {
			out = append(out, f)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func (s *ReviewState) FindingsByStatus(status string) []Finding {
	var out []Finding
	for _, f := range s.Findings {
		if f.Status == status {
			out = append(out, f)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func (s *ReviewState) GetFinding(id int) *Finding {
	fid := fmt.Sprintf("%d", id)
	f, ok := s.Findings[fid]
	if !ok {
		return nil
	}
	return &f
}

func (s *ReviewState) LatestRound() *ReviewRound {
	if len(s.ReviewRounds) == 0 {
		return nil
	}
	return &s.ReviewRounds[len(s.ReviewRounds)-1]
}

func (s *ReviewState) SortedFindingIDs() []int {
	var ids []int
	for _, f := range s.Findings {
		ids = append(ids, f.ID)
	}
	sort.Ints(ids)
	return ids
}

func (s *ReviewState) StateSummaryText() string {
	var b strings.Builder
	sm := s.Summary
	fmt.Fprintf(&b, "## Review State (Round %d)\n", len(s.ReviewRounds))
	fmt.Fprintf(&b, "- Total findings: %d\n", sm.TotalFindings)
	fmt.Fprintf(&b, "- Open: %d\n", sm.Open)
	fmt.Fprintf(&b, "- Resolved: %d\n", sm.Resolved)
	fmt.Fprintf(&b, "- Won't fix: %d\n", sm.Wontfix)
	fmt.Fprintf(&b, "- Retracted: %d\n", sm.Retracted)
	fmt.Fprintf(&b, "- Assessment: %s\n", sm.Assessment)
	sha := sm.LastReviewedSHA
	if len(sha) > 8 {
		sha = sha[:8]
	}
	if sha == "" {
		sha = "none"
	}
	fmt.Fprintf(&b, "- Last reviewed SHA: %s\n", sha)
	b.WriteString("\n### Active Findings:\n")
	for _, f := range s.OpenFindings() {
		fmt.Fprintf(&b, "  - **Finding %d** [%s] (%s): %s — `%s:%d`\n",
			f.ID, f.Severity, f.Status, f.Title, f.File, f.Line)
	}
	resolved := s.FindingsByStatus("resolved")
	if len(resolved) > 0 {
		b.WriteString("\n### Resolved Findings:\n")
		for _, f := range resolved {
			note := ""
			for i := len(f.StatusHistory) - 1; i >= 0; i-- {
				if f.StatusHistory[i].Note != "" {
					note = " — " + f.StatusHistory[i].Note
					break
				}
			}
			fmt.Fprintf(&b, "  - ~~Finding %d~~ [%s]: %s%s\n", f.ID, f.Severity, f.Title, note)
		}
	}
	return b.String()
}

func (s *ReviewState) RecomputeSummary() {
	s.Summary.TotalFindings = len(s.Findings)
	s.Summary.Open = 0
	s.Summary.Resolved = 0
	s.Summary.Wontfix = 0
	s.Summary.Retracted = 0
	hasCritical := false
	for _, f := range s.Findings {
		switch f.Status {
		case "open", "partially_fixed":
			s.Summary.Open++
			if f.Severity == "CRITICAL" {
				hasCritical = true
			}
		case "resolved":
			s.Summary.Resolved++
		case "wontfix":
			s.Summary.Wontfix++
		case "retracted":
			s.Summary.Retracted++
		}
	}
	if s.Summary.Open == 0 && s.Summary.TotalFindings > 0 {
		s.Summary.Assessment = "APPROVE"
	} else if hasCritical || s.Summary.Open > 0 {
		s.Summary.Assessment = "REQUEST CHANGES"
	} else {
		s.Summary.Assessment = "PENDING"
	}
}

// ── GitHub Comment Backend ──────────────────────────────────────────────────

const (
	stateMarkerStart  = "<!-- REVIEWIQ_STATE_START -->"
	stateMarkerEnd    = "<!-- REVIEWIQ_STATE_END -->"
	stateCommentHeader = "<!-- REVIEWIQ_STATE_COMMENT -->"
)

func stateRoundMarker(round int) string {
	return fmt.Sprintf("<!-- REVIEWIQ_STATE_ROUND_%d -->", round)
}

func LoadRemote(repo string, prNumber int) (*ReviewState, error) {
	_, s, err := findLatestStateComment(repo, prNumber)
	return s, err
}

// SaveRemote always creates a NEW comment (never overwrites previous rounds).
// Each round's state is preserved as a separate hidden comment for audit trail.
func SaveRemote(s *ReviewState) error {
	encoded, err := encodeState(s)
	if err != nil {
		return err
	}
	sm := s.Summary
	round := len(s.ReviewRounds)
	body := fmt.Sprintf(`%s
%s
<details>
<summary>ReviewIQ State (Round %d) — %d open, %d resolved</summary>

| Metric | Count |
|--------|-------|
| Total findings | %d |
| Open | %d |
| Resolved | %d |
| Won't fix | %d |
| Retracted | %d |
| Assessment | %s |

</details>

%s
%s
%s`,
		stateCommentHeader, stateRoundMarker(round), round, sm.Open, sm.Resolved,
		sm.TotalFindings, sm.Open, sm.Resolved, sm.Wontfix, sm.Retracted, sm.Assessment,
		stateMarkerStart, encoded, stateMarkerEnd)

	// Always POST new comment — preserves all previous round states
	return ghAPI("POST", fmt.Sprintf("/repos/%s/issues/%d/comments", s.PR.Repo, s.PR.Number), map[string]string{"body": body})
}

// ── Load/Save (GitHub-only) ─────────────────────────────────────────────────

func Load(prNumber int, repo string) *ReviewState {
	if repo != "" {
		s, _ := LoadRemote(repo, prNumber)
		if s != nil {
			return s
		}
	}
	return NewState(prNumber, repo)
}

func Save(s *ReviewState) {
	if s.PR.Repo != "" {
		_ = SaveRemote(s)
	}
}

// ── Helpers ─────────────────────────────────────────────────────────────────

func nowUTC() string { return time.Now().UTC().Format(time.RFC3339) }

func encodeState(s *ReviewState) (string, error) {
	data, err := json.Marshal(s)
	if err != nil {
		return "", err
	}
	return base64.StdEncoding.EncodeToString(data), nil
}

func hasGitHubAuth() bool {
	token, err := github.GetToken()
	return err == nil && token != ""
}

// findLatestStateComment finds all state comments and returns the one with the highest round number.
func findLatestStateComment(repo string, prNumber int) (int, *ReviewState, error) {
	token, err := github.GetToken()
	if err != nil {
		return 0, nil, nil
	}
	url := fmt.Sprintf("https://api.github.com/repos/%s/issues/%d/comments", repo, prNumber)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", "token "+token)
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return 0, nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var comments []struct {
		ID   int    `json:"id"`
		Body string `json:"body"`
	}
	if err := json.Unmarshal(body, &comments); err != nil {
		return 0, nil, nil
	}
	re := regexp.MustCompile(regexp.QuoteMeta(stateMarkerStart) + `\n(.+?)\n` + regexp.QuoteMeta(stateMarkerEnd))

	var bestID int
	var bestState *ReviewState
	bestRound := -1

	for _, c := range comments {
		if !strings.Contains(c.Body, stateCommentHeader) {
			continue
		}
		match := re.FindStringSubmatch(c.Body)
		if len(match) < 2 {
			continue
		}
		decoded, err := base64.StdEncoding.DecodeString(strings.TrimSpace(match[1]))
		if err != nil {
			continue
		}
		var s ReviewState
		if err := json.Unmarshal(decoded, &s); err != nil {
			continue
		}
		round := len(s.ReviewRounds)
		if round > bestRound {
			bestRound = round
			bestID = c.ID
			bestState = &s
		}
	}
	if bestState != nil {
		return bestID, bestState, nil
	}
	return 0, nil, nil
}

func ghAPI(method, endpoint string, data map[string]string) error {
	token, err := github.GetToken()
	if err != nil {
		return err
	}
	body, _ := json.Marshal(data)
	url := "https://api.github.com" + endpoint
	req, _ := http.NewRequest(method, url, strings.NewReader(string(body)))
	req.Header.Set("Authorization", "token "+token)
	req.Header.Set("Accept", "application/vnd.github.v3+json")
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	resp.Body.Close()
	return nil
}
